### ISSUE TITLE
no-shadow doesn't seem to work with const/let

### ISSUE THREAD

**Issue by paulcbetts** [eslint/eslint#3633]
Example:

This reports no error on ESLint v1.3.1:

```js
const foo = "Bar";
let foo = "Baz";
```

This correctly fails with no-const-assign:

```js
const foo = "Bar";
foo = "Baz";
```

My eslintrc file:

```js
{
    "parser": "babel-eslint",
    "rules": {
        "strict": 0,
        "indent": [
            2,
            2
        ],
        "semi": [
            2,
            "always"
        ],
        "arrow-parens": 2,
        "arrow-spacing": 1,
        "constructor-super": 2,
        "no-class-assign": 2,
        "no-const-assign": 2,
        "no-dupe-class-members": 2,
        "no-this-before-super": 2,
        "callback-return": 2,
        "no-path-concat": 2,
        "no-label-var": 2,
        "no-shadow": 2,
        "handle-callback-err": 1,
        "prefer-spread": 1,
        "prefer-arrow-callback": 1,
        "prefer-template": 1,
        "no-console": 0
    },
    "env": {
        "es6": true,
        "node": true,
        "browser": true
    },
    "extends": "eslint:recommended"
}
```

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
Ah, so this is interesting.

```js
const foo = "Bar";
let foo = "Baz";
```

This is technically not shadowing, as shadowing means that a declaration has the same name as an existing binding in an upper scope. In this case, what you have are two declarations for the same binding in the same scope. So really the rule we're talking about here is `no-redeclare`.

From my test, it looks like `no-redeclare` flags this case. Can you confirm?

**Comment by paulcbetts** [comment]
> From my test, it looks like no-redeclare flags this case. Can you confirm?

This doesn't seem to flag for me, but maybe I'm running an out-of-date version or have something borked in my setup. Another case that should be flagged is paving over imports:

```js
import fs from 'fs';

fs = 4;
// or
let fs = 72;
```

IMHO, import statements should be treated as const, though I don't know where the ES6 spec lands on this - I imagine that 95%+ of the time, reassigning an import is done accidentally

**Comment by ilyavolodin** [comment]
Imports should be although caught under `no-redeclare`, just verified in online demo and `let fs = 72` will be flagged.

**Comment by paulcbetts** [comment]
Weird, it's like no-redeclare has no effect on my box. Hm. Well, I'll consider this closed since it seems like I'm Doing Something Wrong™

**Comment by ilyavolodin** [comment]
Let us know if you will not be able to figure out what's causing your issue.

**Comment by michaelficarra** [comment]
No rules should be handling these because they are not programs. If they are parsing, an issue should be opened in the parser that you are using.

**Comment by ilyavolodin** [comment]
@michaelficarra Why would this be a parser issue? It seems like legal code to me that you shouldn't write.

**Comment by michaelficarra** [comment]
[15.1.1](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-scripts-static-semantics-early-errors):

> * It is a Syntax Error if the LexicallyDeclaredNames of ScriptBody contains any duplicate entries.
> * It is a Syntax Error if any element of the LexicallyDeclaredNames of ScriptBody also occurs in the VarDeclaredNames of ScriptBody.

[15.2.1.1](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-module-semantics-static-semantics-early-errors):

> * It is a Syntax Error if the LexicallyDeclaredNames of ModuleItemList contains any duplicate entries.
> * It is a Syntax Error if any element of the LexicallyDeclaredNames of ModuleItemList also occurs in the VarDeclaredNames of ModuleItemList.

[13.2.1](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-block-static-semantics-early-errors):

> * It is a Syntax Error if the LexicallyDeclaredNames of StatementList contains any duplicate entries.
> * It is a Syntax Error if any element of the LexicallyDeclaredNames of StatementList also occurs in the VarDeclaredNames of StatementList.

... etc.

**Comment by ilyavolodin** [comment]
Nice! Thanks for the explanation. In that case, yes, this should be an espree ticket. I wonder if Esprima catches this, or acorn.

**Comment by paulcbetts** [comment]
Aha, so I'm using babel-eslint which is probably letting this slide

**Comment by michaelficarra** [comment]
The Shift parser gets it right :grin:.

**Comment by ilyavolodin** [comment]
Esprima also doesn't get it right. Can't find online demo for acorn to test this out.
